File: aoc-py/solutions/day2.py
"""
Day 2: Red-Nosed Reports

https://adventofcode.com/2024/day/2
"""
__all__ = ('Day2',)

from typing import ClassVar

from ..solution import Solution

class Day2(Solution):
    NAME: ClassVar[str] = 'Red-Nosed Reports'

    def part_one(self, inp: str) -> int:
        count = 0

        for line in inp.splitlines():
            report = [int(level) for level in line.split()]
            increasing = True

            for i in range(len(report) - 1):
                jump = report[i + 1] - report[i]

                if i == 0:
                    increasing = jump > 0
                if (jump if increasing else -jump) not in range(1, 4):
                    break
            else:
                count += 1
        return count

    def part_two(self, inp: str) -> int:
        count = 0

        for line in inp.splitlines():
            report = [int(level) for level in line.split()]

            n = len(report)

            # test all possible levels to remove
            for remove in range(n):
                new_report = report[:remove] + report[remove + 1:]
                increasing = True

                for i in range(n - 2):
                    jump = new_report[i + 1] - new_report[i]

                    if i == 0:
                        increasing = jump > 0
                    if (jump if increasing else -jump) not in range(1, 4):
                        break
                else:
                    count += 1
                    break
        return count

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 585
        assert p2 == 626

File: src/bin/day2.rs
//! Day 2: Red-Nosed Reports
//!
//! <https://adventofcode.com/2024/day/2>
#![feature(array_windows)]

use std::fmt::Display;
use aoc_2024::Solution;

pub struct Day2;

impl Solution for Day2 {
    const NAME: &'static str = "Red-Nosed Reports";

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        inp.to_string()
            .lines()
            .filter(|line| {
                let report = line
                    .split_whitespace()
                    .filter_map(|x| x.parse::<isize>().ok())
                    .collect::<Vec<_>>();
                let mut increasing = false;

                !report.array_windows::<2>()
                    .enumerate()
                    .any(|(i, [a, b])| {
                        let jump = b - a;

                        if i == 0 {
                            increasing = jump > 0;
                        }

                        !(1..=3).contains(&if increasing { jump } else { -jump })
                    })
            })
            .count()
    }

    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        inp.to_string()
            .lines()
            .filter(|line| {
                (0..line.len()).any(|remove| {
                    let report = line
                        .split_whitespace()
                        .enumerate()
                        .filter_map(|(i, x)| x.parse::<isize>().ok().filter(|_| i != remove))
                        .collect::<Vec<_>>();
                    let mut increasing = false;

                    !report.array_windows::<2>()
                        .enumerate()
                        .any(|(i, [a, b])| {
                            let jump = b - a;

                            if i == 0 {
                                increasing = jump > 0;
                            }

                            !(1..=3).contains(&if increasing { jump } else { -jump })
                        })
                })
            })
            .count()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 585);
        assert_eq!(p2, 626);
    }
}

fn main() {
    aoc_2024::run_day(2, &Day2);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}